### ISSUE TITLE
Relative Uri path has an empty item as head

### ISSUE THREAD

**Issue by kgston** [softwaremill/sttp-model#59]
Thanks for the quick fix in 1.2.1! 
I bumped into this (maybe cosmetic?) issue where the `path` list of path segments has a blank entry as the first object when the uri is in the form `/a/b`. If you tried to combine the relative path with another absolute path, you would end up with a resulting Uri with `//` where it was concatenated. It is fine for the case where the Uri is parsed from `a/b`, but if the relative Uri only has a single path segment like `a`, it would not parse properly without the leading `/` denoting it is a relative path.

```scala
import sttp.model._

val test = Uri.unsafeParse("/a")
val endpoint = Uri.unsafeParse("http://test.com/abc")
println(test.path) // List(, a)
println(endpoint.addPath(test.path)) // http://test.com/abc//a
```

**Comment by adamw** [comment]
Hm good catch and good question!

The problem here is that we need to both represent path-relative (e.g. `a/b/c`) and path-absolute (e.g. `/a/b/c`) relative URIs. WIth absolute URIs this is simple, as there are no relative paths. Here we've got this additional complexity.

An open question is how to solve this :). We've need to somehow let the user know (at least through documentation) what to expect from `.path` and how to recognize path-absolute and path-relative URIs.

One way is introducing a `pathAbsolute: Boolean` flag to the `URI` class. But then ... wouldn't it be surprising that both `Uri.unsafeParse("/a").path` and `Uri.unsafeParse("a").path` return the same? The paths are different in both cases.

I'm open to suggestions and ideas :)

**Comment by adamw** [comment]
One more note: probably the example is just a demo of the problem, but if you've got a base URI & want to make it more specific, I think either replacing the path with `.withPath` or `.resolve` could work better?

Adding an absolute path to an URI with a path already defined - not sure what the result here should be :)

**Comment by kgston** [comment]
Hmm, I didn't know about the difference in the path-relative & path-absolute implementation. I could work with just the path-relative format, except that I also noticed that paths with a single segment like `Uri.unsafeParse("path1").path` would be `List()` and `path1` would end up being parsed as the `scheme`. Not sure if this is a bug or intended, but if single segment relative path can be done, it would fit my use case.

One use case that I was working on was with AWS API Gateway. They have this format where it ends with a path segment as part of the base URI. For example `https://123456789.execute-api.ap-northeast-1.amazonaws.com/main`. So I would have 2 instances of that for different environments, e.g.
- `https://123456789.execute-api.ap-northeast-1.amazonaws.com/main` - for test environment
- `https://abcdefg.execute-api.ap-northeast-1.amazonaws.com/main` - for prod environment

but same relative path `/path/to/my/api` to the resource, hence the need for using `addPath`. The end result here would be like `https://123456789.execute-api.ap-northeast-1.amazonaws.com/main/path/to/my/api`.

Of course one could make the case that then the path should be `/main/path/to/my/api`, but `/main` path segment is arguably an "infra" concern that should be abstracted away from the code base since only `path/to/my/api` is what is being declared by the application. If we factor on custom domain names for AWS API Gateway, the final endpoint will become `https://my.domain.example/path/to/my/api`, with the AWS generated domain and `/main` being abstracted out.

I also tried out `.resolve` and that seems to take into account the difference between path-relative & path-absolute, so it would not work for single segment relative path yet.

**Comment by adamw** [comment]
Forgot to link the commit: https://github.com/softwaremill/sttp-model/commit/550d2fa09b7262923e6320fd1fa2aedf6f43e721

**Comment by adamw** [comment]
I've fixed the single-path-segment uri parsing in 1.2.2. I think you should be able to properly parse relative-uris with relative-paths, and hence also properly construct the target URIs.

**Comment by kgston** [comment]
Thanks! The single-path-segment uri parsing works and is good enough for my use case.

I'm assuming that the empty head is used to denote the difference between path-relative & path-absolute? If so, you may want to consider wrapping (and maybe even extending from an abstract ADT) the path segment list in an ADT. It'll be easy to pattern match on specific subtypes or to unapply the ADT for its contents. That way there won't be an empty head and no need for OO style flags.

**Comment by adamw** [comment]
Yes, that's an option, but brings some questions ;) How about this design:

1. what should be the default, that is when we start with an empty URI - I guess we would start with an absolute-path (in the otherwise empty, hence relative URI). 
2. the `.addPath` methods would add segments to whatever path the URI has currently. So in the default case - the absolute path, but if it's relative, we would add segmetns to the relative one
3. `.withPath` would always replace the path with an absolute one
4. we would need a separate `.withRelativePath` which would replace the path and set it to a relative one

**Comment by kgston** [comment]
1. I think defaulting to an absolute path with an empty URI might make sense. Alternatively, we could have an ADT representing `NoPath`? Might be overkill, but I think it opens up some possibilities.
2. I think if there is an abstract ADT like `abstract class Path(path: List[String])` (could also be a trait), `.addPath` could be implemented on that level instead of the concrete ADT (e.g. `AbsolutePath`, `RelativePath`, `NoPath`), and `.addPath` on URI would link to that - but I agree with the general idea.
3. I think `.withPath(string: String)` should just replace the contents of the existing path with the given path, but not change its type. I think having helper methods like `.asRelativePath` & `.asAbsolutePath` would make sense when changing between path types.
Another suggestion would be to have `withPath(path: Path)` and have the developer choose the concrete ADT type.
4. This would be solved with the suggestion in 3.

**Comment by adamw** [comment]
Released in 1.3.1
